### ISSUE TITLE
Queries with `having` result in QScript with multiple joins

### ISSUE THREAD

**Issue by jsantos17** [precog/quasar#2932]
The query:

`select city, sum(pop) from zips group by city having sum(pop) > 50000`

Compiling the query yields the following, immense QScript.  For Mongo, the QScript is simplified somewhat but does not execute with `value too large to reduce`. It also requires several `mapReduce` phases. No integration tests use `having` at the moment.

<details>
  <summary>Full QScript with four `ThetaJoin`s</summary>
 

```
QScript:
ThetaJoin
├─ ThetaJoin
│  ├─ LeftShift
│  │  ├─ Read(/mngo/jps/zips)
│  │  ├─ Struct
│  │  │  ╰─ ○
│  │  ├─ IdStatus
│  │  │  ╰─ IdStatus(IncludeId)
│  │  ╰─ Repair
│  │     ╰─ ConcatArrays
│  │        ├─ MakeArray
│  │        │  ╰─ ConcatArrays
│  │        │     ├─ ConcatArrays
│  │        │     │  ├─ ConcatArrays
│  │        │     │  │  ├─ MakeArray
│  │        │     │  │  │  ╰─ MakeArray
│  │        │     │  │  │     ╰─ ProjectIndex
│  │        │     │  │  │        ├─ RightSide
│  │        │     │  │  │        ╰─ Constant(Int(0))
│  │        │     │  │  ╰─ MakeArray
│  │        │     │  │     ╰─ MakeArray
│  │        │     │  │        ╰─ ProjectIndex
│  │        │     │  │           ├─ RightSide
│  │        │     │  │           ╰─ Constant(Int(0))
│  │        │     │  ╰─ MakeArray
│  │        │     │     ╰─ Guard
│  │        │     │        ├─ ProjectIndex
│  │        │     │        │  ├─ RightSide
│  │        │     │        │  ╰─ Constant(Int(1))
│  │        │     │        ├─ Type(Obj(Map[], Some(Top)))
│  │        │     │        ├─ ProjectIndex
│  │        │     │        │  ├─ RightSide
│  │        │     │        │  ╰─ Constant(Int(1))
│  │        │     │        ╰─ Undefined
│  │        │     ╰─ MakeArray
│  │        │        ╰─ MakeArray
│  │        │           ╰─ ProjectField
│  │        │              ├─ Guard
│  │        │              │  ├─ ProjectIndex
│  │        │              │  │  ├─ RightSide
│  │        │              │  │  ╰─ Constant(Int(1))
│  │        │              │  ├─ Type(Obj(Map[], Some(Top)))
│  │        │              │  ├─ ProjectIndex
│  │        │              │  │  ├─ RightSide
│  │        │              │  │  ╰─ Constant(Int(1))
│  │        │              │  ╰─ Undefined
│  │        │              ╰─ Constant(Str(city))
│  │        ╰─ MakeArray
│  │           ╰─ ProjectIndex
│  │              ├─ RightSide
│  │              ╰─ Constant(Int(1))
│  ├─ Map
│  │  ├─ ○
│  │  ╰─ ProjectIndex
│  │     ├─ ○
│  │     ╰─ Constant(Int(0))
│  ├─ Reduce
│  │  ├─ ○
│  │  ├─ Bucket
│  │  │  ╰─ MakeArray
│  │  │     ╰─ MakeArray
│  │  │        ╰─ ProjectField
│  │  │           ├─ Guard
│  │  │           │  ├─ ProjectIndex
│  │  │           │  │  ├─ ○
│  │  │           │  │  ╰─ Constant(Int(1))
│  │  │           │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │           │  ├─ ProjectIndex
│  │  │           │  │  ├─ ○
│  │  │           │  │  ╰─ Constant(Int(1))
│  │  │           │  ╰─ Undefined
│  │  │           ╰─ Constant(Str(city))
│  │  ├─ Reducers
│  │  │  ╰─ Sum
│  │  │     ╰─ Guard
│  │  │        ├─ ProjectField
│  │  │        │  ├─ Guard
│  │  │        │  │  ├─ ProjectIndex
│  │  │        │  │  │  ├─ ○
│  │  │        │  │  │  ╰─ Constant(Int(1))
│  │  │        │  │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │        │  │  ├─ ProjectIndex
│  │  │        │  │  │  ├─ ○
│  │  │        │  │  │  ╰─ Constant(Int(1))
│  │  │        │  │  ╰─ Undefined
│  │  │        │  ╰─ Constant(Str(pop))
│  │  │        ├─ Type(Int, Dec, or Interval)
│  │  │        ├─ ProjectField
│  │  │        │  ├─ Guard
│  │  │        │  │  ├─ ProjectIndex
│  │  │        │  │  │  ├─ ○
│  │  │        │  │  │  ╰─ Constant(Int(1))
│  │  │        │  │  ├─ Type(Obj(Map[], Some(Top)))
│  │  │        │  │  ├─ ProjectIndex
│  │  │        │  │  │  ├─ ○
│  │  │        │  │  │  ╰─ Constant(Int(1))
│  │  │        │  │  ╰─ Undefined
│  │  │        │  ╰─ Constant(Str(pop))
│  │  │        ╰─ Undefined
│  │  ╰─ Repair
│  │     ╰─ ConcatArrays
│  │        ├─ MakeArray
│  │        │  ╰─ ReduceIndex(None)
│  │        ╰─ MakeArray
│  │           ╰─ ReduceIndex(Some(0))
│  ├─ And
│  │  ├─ Constant(Bool(true))
│  │  ╰─ Gt
│  │     ├─ ProjectIndex
│  │     │  ├─ RightSide
│  │     │  ╰─ Constant(Int(1))
│  │     ╰─ Constant(Int(50000))
│  ├─ Inner
│  ╰─ ConcatArrays
│     ├─ MakeArray
│     │  ╰─ ConcatArrays
│     │     ├─ ConcatArrays
│     │     │  ├─ ConcatArrays
│     │     │  │  ├─ MakeArray
│     │     │  │  │  ╰─ ConcatArrays
│     │     │  │  │     ├─ ConcatArrays
│     │     │  │  │     │  ├─ MakeArray
│     │     │  │  │     │  │  ╰─ ProjectIndex
│     │     │  │  │     │  │     ├─ ProjectIndex
│     │     │  │  │     │  │     │  ├─ LeftSide
│     │     │  │  │     │  │     │  ╰─ Constant(Int(0))
│     │     │  │  │     │  │     ╰─ Constant(Int(0))
│     │     │  │  │     │  ╰─ MakeArray
│     │     │  │  │     │     ╰─ ProjectIndex
│     │     │  │  │     │        ├─ ProjectIndex
│     │     │  │  │     │        │  ├─ LeftSide
│     │     │  │  │     │        │  ╰─ Constant(Int(1))
│     │     │  │  │     │        ╰─ Constant(Int(0))
│     │     │  │  │     ╰─ MakeArray
│     │     │  │  │        ╰─ ProjectIndex
│     │     │  │  │           ├─ LeftSide
│     │     │  │  │           ╰─ Constant(Int(3))
│     │     │  │  ╰─ MakeArray
│     │     │  │     ╰─ MakeArray
│     │     │  │        ╰─ ProjectIndex
│     │     │  │           ├─ ProjectIndex
│     │     │  │           │  ├─ RightSide
│     │     │  │           │  ╰─ Constant(Int(0))
│     │     │  │           ╰─ Constant(Int(0))
│     │     │  ╰─ MakeArray
│     │     │     ╰─ ProjectIndex
│     │     │        ├─ LeftSide
│     │     │        ╰─ Constant(Int(2))
│     │     ╰─ MakeArray
│     │        ╰─ Gt
│     │           ├─ ProjectIndex
│     │           │  ├─ RightSide
│     │           │  ╰─ Constant(Int(1))
│     │           ╰─ Constant(Int(50000))
│     ╰─ MakeArray
│        ╰─ ConcatArrays
│           ├─ ConcatArrays
│           │  ├─ ConcatArrays
│           │  │  ├─ MakeArray
│           │  │  │  ╰─ ConcatArrays
│           │  │  │     ├─ ConcatArrays
│           │  │  │     │  ├─ ConcatArrays
│           │  │  │     │  │  ├─ MakeArray
│           │  │  │     │  │  │  ╰─ ProjectIndex
│           │  │  │     │  │  │     ├─ ProjectIndex
│           │  │  │     │  │  │     │  ├─ LeftSide
│           │  │  │     │  │  │     │  ╰─ Constant(Int(0))
│           │  │  │     │  │  │     ╰─ Constant(Int(0))
│           │  │  │     │  │  ╰─ MakeArray
│           │  │  │     │  │     ╰─ ProjectIndex
│           │  │  │     │  │        ├─ ProjectIndex
│           │  │  │     │  │        │  ├─ LeftSide
│           │  │  │     │  │        │  ╰─ Constant(Int(1))
│           │  │  │     │  │        ╰─ Constant(Int(0))
│           │  │  │     │  ╰─ MakeArray
│           │  │  │     │     ╰─ ProjectIndex
│           │  │  │     │        ├─ LeftSide
│           │  │  │     │        ╰─ Constant(Int(3))
│           │  │  │     ╰─ MakeArray
│           │  │  │        ╰─ ProjectIndex
│           │  │  │           ├─ ProjectIndex
│           │  │  │           │  ├─ RightSide
│           │  │  │           │  ╰─ Constant(Int(0))
│           │  │  │           ╰─ Constant(Int(0))
│           │  │  ╰─ MakeArray
│           │  │     ╰─ ConcatArrays
│           │  │        ├─ ConcatArrays
│           │  │        │  ├─ ConcatArrays
│           │  │        │  │  ├─ MakeArray
│           │  │        │  │  │  ╰─ ProjectIndex
│           │  │        │  │  │     ├─ ProjectIndex
│           │  │        │  │  │     │  ├─ LeftSide
│           │  │        │  │  │     │  ╰─ Constant(Int(0))
│           │  │        │  │  │     ╰─ Constant(Int(0))
│           │  │        │  │  ╰─ MakeArray
│           │  │        │  │     ╰─ ProjectIndex
│           │  │        │  │        ├─ ProjectIndex
│           │  │        │  │        │  ├─ LeftSide
│           │  │        │  │        │  ╰─ Constant(Int(1))
│           │  │        │  │        ╰─ Constant(Int(0))
│           │  │        │  ╰─ MakeArray
│           │  │        │     ╰─ ProjectIndex
│           │  │        │        ├─ LeftSide
│           │  │        │        ╰─ Constant(Int(3))
│           │  │        ╰─ MakeArray
│           │  │           ╰─ ProjectIndex
│           │  │              ├─ ProjectIndex
│           │  │              │  ├─ RightSide
│           │  │              │  ╰─ Constant(Int(0))
│           │  │              ╰─ Constant(Int(0))
│           │  ╰─ MakeArray
│           │     ╰─ ProjectField
│           │        ├─ ProjectIndex
│           │        │  ├─ LeftSide
│           │        │  ╰─ Constant(Int(2))
│           │        ╰─ Constant(Str(pop))
│           ╰─ MakeArray
│              ╰─ ProjectField
│                 ├─ ProjectIndex
│                 │  ├─ LeftSide
│                 │  ╰─ Constant(Int(2))
│                 ╰─ Constant(Str(pop))
├─ Reduce
│  ├─ ○
│  ├─ Bucket
│  │  ╰─ ConcatArrays
│  │     ├─ MakeArray
│  │     │  ╰─ ProjectIndex
│  │     │     ├─ ProjectIndex
│  │     │     │  ├─ ProjectIndex
│  │     │     │  │  ├─ ○
│  │     │     │  │  ╰─ Constant(Int(0))
│  │     │     │  ╰─ Constant(Int(0))
│  │     │     ╰─ Constant(Int(2))
│  │     ╰─ MakeArray
│  │        ╰─ ProjectIndex
│  │           ├─ ProjectIndex
│  │           │  ├─ ProjectIndex
│  │           │  │  ├─ ○
│  │           │  │  ╰─ Constant(Int(0))
│  │           │  ╰─ Constant(Int(1))
│  │           ╰─ Constant(Int(0))
│  ├─ Reducers
│  │  ╰─ Arbitrary
│  │     ╰─ ProjectField
│  │        ├─ ProjectIndex
│  │        │  ├─ ProjectIndex
│  │        │  │  ├─ ○
│  │        │  │  ╰─ Constant(Int(0))
│  │        │  ╰─ Constant(Int(2))
│  │        ╰─ Constant(Str(city))
│  ╰─ Repair
│     ╰─ ConcatArrays
│        ├─ MakeArray
│        │  ╰─ ReduceIndex(None)
│        ╰─ MakeArray
│           ╰─ ReduceIndex(Some(0))
├─ Reduce
│  ├─ ○
│  ├─ Bucket
│  │  ╰─ ConcatArrays
│  │     ├─ MakeArray
│  │     │  ╰─ ProjectIndex
│  │     │     ├─ ProjectIndex
│  │     │     │  ├─ ProjectIndex
│  │     │     │  │  ├─ ○
│  │     │     │  │  ╰─ Constant(Int(1))
│  │     │     │  ╰─ Constant(Int(0))
│  │     │     ╰─ Constant(Int(2))
│  │     ╰─ MakeArray
│  │        ╰─ ProjectIndex
│  │           ├─ ProjectIndex
│  │           │  ├─ ProjectIndex
│  │           │  │  ├─ ○
│  │           │  │  ╰─ Constant(Int(1))
│  │           │  ╰─ Constant(Int(0))
│  │           ╰─ Constant(Int(3))
│  ├─ Reducers
│  │  ╰─ Sum
│  │     ╰─ Guard
│  │        ├─ ProjectIndex
│  │        │  ├─ ProjectIndex
│  │        │  │  ├─ ○
│  │        │  │  ╰─ Constant(Int(1))
│  │        │  ╰─ Constant(Int(2))
│  │        ├─ Type(Int, Dec, or Interval)
│  │        ├─ ProjectIndex
│  │        │  ├─ ProjectIndex
│  │        │  │  ├─ ○
│  │        │  │  ╰─ Constant(Int(1))
│  │        │  ╰─ Constant(Int(3))
│  │        ╰─ Undefined
│  ╰─ Repair
│     ╰─ ConcatArrays
│        ├─ MakeArray
│        │  ╰─ ReduceIndex(None)
│        ╰─ MakeArray
│           ╰─ ReduceIndex(Some(0))
├─ Constant(Bool(true))
├─ Inner
╰─ ConcatMaps
   ├─ MakeMap
   │  ├─ Constant(Str(city))
   │  ╰─ ProjectIndex
   │     ├─ LeftSide
   │     ╰─ Constant(Int(1))
   ╰─ MakeMap
      ├─ Constant(Str(1))
      ╰─ ProjectIndex
         ├─ RightSide
         ╰─ Constant(Int(1))
```

</details>

**Comment by alissapajer** [comment]
This may be related to a qscript provenance issue, discovered in debugging #1929. qscript doesn't properly align join provenance with other related provenances. So if there are multiple joins in the qscript, then it is likely that they are a result of mergeable sides not merging due to this bug.

**Comment by alissapajer** [comment]
I also want to see what improved `DeepShape` merging can bring to this. We don't currently deep merge through `FreeQS` branches, but I'm currently working on doing that for another issue. So I'll test this query when I'm done with that.

**Comment by djspiewak** [comment]
I'm going to add this to the QSU epic, since it seems intrinsically related.  Not skulling though.

**Comment by djspiewak** [comment]
Testing this is blocked on #3021.  I expect it is fixed though.

**Comment by djspiewak** [comment]
Actually, looking at this more closely…  This is *exactly* #3021 and will be fixed by it.

**Comment by alissapajer** [comment]
@djspiewak as of 27.2.0 this is hitting the `ReifyBuckets: Expected to find a mappable region` error. The part of #3021 that fixes this (handling nonmappable filter predicates) is already in master.

**Comment by alissapajer** [comment]
@wemrysi is this something that will be fixed by your current work?

**Comment by wemrysi** [comment]
@alissapajer I don't _think_ so, but the LP generated for this query looks fishy to me. It appears we're filtering the grouped dataset by the sum > 50,000 and then reducing the result of that _again_, which doesn't seem correct. Would you concur, or have I misinterpreted this?

<details>
  <summary>Optimized LP</summary>

```
Let('__tmp0)                                                                  
├─ Read(/zips)                           
╰─ Let('__tmp1)                                                      
   ├─ Invoke(GroupBy)                 
   │  ├─ Free('__tmp0)            
   │  ╰─ Invoke(MakeArray)                                                                               
   │     ╰─ MapProject(Free('__tmp0){"city"})
   ╰─ Let('__tmp2)                                                 
      ├─ Invoke(Filter)                                               
      │  ├─ Free('__tmp1)
      │  ╰─ Invoke(Gt)    
      │     ├─ Invoke(Sum)                   
      │     │  ╰─ MapProject(Free('__tmp1){"pop"})                                                                                
      │     ╰─ Constant(Int(50000))                                                
      ╰─ Invoke(Squash)                                                            
         ╰─ Invoke(MapConcat)                                                                                                 
            ├─ Invoke(MakeMap)                   
            │  ├─ Str("city") 
            │  ╰─ Invoke(Arbitrary)                  
            │     ╰─ MapProject(Free('__tmp2){"city"})                                                                            
            ╰─ Invoke(MakeMap)                                                     
               ├─ Str("1")           
               ╰─ Invoke(Sum)       
                  ╰─ MapProject(Free('__tmp2){"pop"})
```
</details>

**Comment by djspiewak** [comment]
Oh yeah that’s definitely weird. I concur that the LP is wrong.

**Comment by alissapajer** [comment]
Agreed. 

The original issue was for qscript with too many joins. Should we open a new issue to track the LP bug, and then keep this issue open, coming back to it after the LP bug is fixed?

**Comment by djspiewak** [comment]
Let’s close this and open a new one for the LP.

**Comment by djspiewak** [comment]
Oh you’re right nvm

**Comment by alissapajer** [comment]
Yeah, this is a unique case with the reduction appearing in the selection and grouping, so I think it's worth checking the qscript (and making an integration test for it) after the LP is correct.

**Comment by djspiewak** [comment]
Agreed. We can unskull this though
